src/test/java/org/javarosa/core/model/DynamicSelectUpdateTest.java

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright (C) 2021 ODK
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.javarosa.core.model;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.sameInstance;
+import static org.javarosa.core.test.SelectChoiceMatchers.choice;
+import static org.javarosa.core.util.XFormsElement.body;
+import static org.javarosa.core.util.XFormsElement.head;
+import static org.javarosa.core.util.XFormsElement.html;
+import static org.javarosa.core.util.XFormsElement.input;
+import static org.javarosa.core.util.XFormsElement.instance;
+import static org.javarosa.core.util.XFormsElement.item;
+import static org.javarosa.core.util.XFormsElement.mainInstance;
+import static org.javarosa.core.util.XFormsElement.model;
+import static org.javarosa.core.util.XFormsElement.repeat;
+import static org.javarosa.core.util.XFormsElement.select1Dynamic;
+import static org.javarosa.core.util.XFormsElement.t;
+import static org.javarosa.core.util.XFormsElement.title;
+
+import java.util.List;
+import org.javarosa.core.test.Scenario;
+import org.javarosa.core.util.XFormsElement;
+import org.junit.Test;
+
+/**
+ * Integration tests to verify that the choice lists for "dynamic selects" (selects with itemsets rather than inline items) are updated when
+ * dependent values change.
+ *
+ * See also:
+ * - {@see SelectOneChoiceFilterTest}
+ * - {@see SelectMultipleChoiceFilterTest} for coverage of dynamic select multiples
+ * - {@see XPathFuncExprRandomizeTest} for coverage of choice list updates when randomization is specified
+ */
+public class DynamicSelectUpdateTest {
```

Comment:
<img width="1082" alt="image" src="https://user-images.githubusercontent.com/55099/111084907-01047500-84d2-11eb-874b-900695e2e75e.png">
Should we have a test case where the cache is not used because the random seed differs? (The picture shows code coverage.)

Reply:
Did you see that the tests in `XPathFuncExprRandomizeTest` do exercise that case? I ended up putting a reference to that test class in the header comment for `DynamicSelectUpdateTest`. Do you think it's important to have explicit tests in both places?

Reply:
Oh, sorry, I missed that!